Meta: Add templates/checklists for new, Accepted/Rejected & Final PEPs (#2956)
python/peps: .github/PULL_REQUEST_TEMPLATE/new-pep.md

---
name: New PEP
about: Submit a new draft PEP

> title: PEP NNN: Title of PEP

Reviewer: ```suggestion
title: "PEP NNN: Title of PEP"
```

---

> name: New PEP
> about: Submit a new draft PEP for consideration

Author: ```suggestion
name: New Standards Track PEP
about: Submit a new draft Standards Track PEP for consideration
```

Adopted the above, using consistent Title Case as in the canonical Type header as opposed to a rather unusual (in English) mid-sentence capitalization of only Standard

> about: Submit a new draft PEP for consideration

Author: Good point, though a lot of it is still fully applicable—presumably, we should make another (shorter/more general) checklist for non-standards-track PEPs with the relevant elements. I can take a crack at it.

Author: After experimenting with several options—including separate Process/Information PEP checklist, and separate Process and Information checklists—I ended up re-organizing the existing new PEP checklist into a section that applied to all PEPs, and a section specific to Standards Track PEPs (that authors were instructed to remove if not applicable). I also caught a few other places (e.g. in the two other templates) where things only apply to Standards Track PEPs, and clarified them accordingly, as well as some other tweaks.

labels: "new-pep"
---

<!--

Please read and follow PEP 1 and PEP 12 before submitting a new PEP:

https://peps.python.org/pep-0001/#submitting-a-pep

https://peps.python.org/pep-0012/#how-to-use-this-template

Make sure to include the PEP number in the pull request title; for example:

PEP 9999: <Title of PEP>

See our Contributing Guidelines (CONTRIBUTING.rst) for more information.

Thanks!

-->

# New PEP

> 
> <!--
> 
> Briefly describe your new PEP here, and anything else we should know.
> Be sure to link to any previous discussions or issues motivating/drafting it.
> 
> -->
> 
> 
> 
> ## Checklist
> 
> <!--
> 
> You should use the following as a guide when double-checking your PEP,
> and you can help complete some of it yourself if you like
> by ticking any boxes you're sure about, like this: [x]
> 
> If you're unsure about anything, just leave it blank and we'll take a look.
> 
> -->
> 

Reviewer 2: ```suggestion

<!--

Resist the temptation to describe your PEP here.
Your PEP's Abstract and Post-History should speak for itself.

You can use the following checklist when double-checking your PEP,
and you can help complete some of it yourself if you like
by ticking any boxes you're sure about, like this: [x]
If you're unsure about anything, just leave it blank and we'll take a look.
-->

```

Author: ```suggestion

<!--

You can use the following checklist when double-checking your PEP,
and you can help complete some of it yourself if you like
by ticking any boxes you're sure about, like this: [x]
If you're unsure about anything, just leave it blank and we'll take a look.

-->

```

The above, minus the extra "resist the temptation" text—this hasn't ever really been an actual problem in the past, and the more structured nature of this template should likely further dissuade if it is. Additionally, a brief meta-description with some additional contextually-relevant details doesn't exactly _hurt_, if authors feel the need to provide it.

* [ ] File created from the [latest PEP template](https://github.com/python/peps/blob/main/pep-0012/pep-NNNN.rst?plain=1)

### Content requirements

> * [ ] PEP topic [discussed in a suitable venue](https://peps.python.org/pep-0001/#start-with-an-idea-for-python) and general agreement that a PEP is appropriate

Author: That's a good practice and should be at least strongly encouraged if not required, but I'm not sure whether it is critical to explicitly state such in this checklist item (as opposed to, e.g., the PEP 1/12 mention of Post-History), given Post-History is mentioned below.

* [ ] Content is reasonably sound, complete and makes at least basic technical sense
* [ ] Abstract and Copyright sections completed, with the required wording in the latter

> * [ ] [Suggested sections](https://peps.python.org/pep-0012/#suggested-sections) included (unless not applicable)

Author: Yeah, I'd certainly find that useful as both a PEP author and reviewer, I was just worried that it would make it too long, but I've gone and and implemented it in another comment

> * [ ] Abstract and Copyright sections completed, with the required wording in the latter
> * [ ] [Suggested sections](https://peps.python.org/pep-0012/#suggested-sections) included (unless not applicable)

Author: ```suggestion
* [ ] [Required and suggested sections](https://peps.python.org/pep-0012/#suggested-sections) included (unless not applicable)
    * [ ] Abstract (required; first section)~
    * [ ] Motivation
    * [ ] Rationale
    * [ ] Specification
    * [ ] Backwards Compatibility
    * [ ] Security Implications
    * [ ] How to Teach This
    * [ ] Reference Implementation
    * [ ] Rejected Ideas
    * [ ] Open Issues
    * [ ] Copyright (required with exact wording; last section)
```

As [suggested by @encukou above](https://github.com/python/peps/pull/2956/files#r1065559112)

> * [ ] Title clearly, accurately and concisely describes the content
> * [ ] Prose is intelligible and generally free of basic grammar and syntax errors, at least to the degree they significantly detract from understanding

Author: ```suggestion
* [ ] Title clearly, accurately and concisely describes the content in 79 characters or less
```

IMO, it's worth mentioning this one explicitly, since it's a header field, has a soft length limit and needs special attention by authors and reviewers (indeed, on the most recent new PEP I reviewed, PEP 702, had a title in need of much adjustment to meet this standard, and this checklist was specifically useful as a reminder of such).

> * [ ] Code is in code blocks, well-formatted (PEP 7/PEP 8) and has the right [lexer name](https://pygments.org/docs/lexers/) if non-Python (or ``text``)

Reviewer 2: This could use a Python-specific section in the devguide. The Pygments docs are too broad.

Author: We already [discuss the specific usage of code blocks and the lexer name](https://peps.python.org/pep-0012/#literal-blocks) in PEP 12; I'll go ahead and link that there—good idea.

```suggestion
* [ ] Code is well-formatted (PEP 7/PEP 8) and is in [code blocks, with the right lexer name](https://peps.python.org/pep-0012/#literal-blocks) if non-Python
```

* [ ] Any project stated in the PEP as supporting/endorsing benefiting from it confirms such

### Technical requirements

* [ ] Filename in the correct format (``pep-NNNN.rst``)
* [ ] ``PEP``, ``Title``, ``Author``, ``Status`` (``Draft``), ``Type`` and ``Created`` headers filled out correctly

> * [ ] ``PEP-Delegate``, ``Requires``, ``Python-Version`` and ``Replaces`` filled out if appropriate
> * [ ] ``Sponsor`` listed and approves if not authored by a core dev or PEP editor

Author: ```suggestion
* [ ] ``Sponsor``, ``PEP-Delegate``, ``Requires``, ``Python-Version`` and ``Replaces`` completed if appropriate
```

Fixed order to match headers, added missing "and" and fixed (original) wording

> * [ ] PEP builds with no warnings, CI checks pass and content displays as intended in the rendered preview

Reviewer 3: This is hard to check ex-ante, given the preview only starts when the PR is opened.

Author: Well, they can always build it locally, check the preview and run the linting before they submit their final PR—that's fully equivalent to the CI. And as mentioned completion any item of the checklist prior to PR submission is entirely optional (though suggested).

Author: I revisied this to change "CI checks" to "pre-commit checks" and "rendered preview" to "rendered HTML", to make it apply equally to local and CI builds.

Author: Marking this as addressed; feel free to re-open or add another comment with any further feedback.

> * [ ] PEP number assigned by the PEP editors and filename & ``PEP`` header updated accordingly

Reviewer 3: Again I think this needs to be clearer that assignment only happens after the PR is opened (e.g. a section *for PEP Editor use*).

Author: I revised the language to reflect that the PEP should be given the next available number, rather than necessarily requiring assignment by the PEP editors. I also was able to consolidate the filename item into this one, and fixed up the grammar/syntax a bit.

Author: Marking this as addressed; feel free to re-open or add another comment with any further feedback.
